### ISSUE TITLE
Add MagicDNS cache in front of forwarder

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#3507]
In a number of configurations, 100.100.100.100 is the platform's sole DNS server and there's no OS-level caching in front of it.

In such cases (e.g. Linux in DNS directMode) we should add DNS caching to the 100.100.100.100 forwarder.

Also: the tsdial.Dialer.UserDial's name resolution as used by SOCKS5 + HTTP outbound proxy when an exit node is in user. There's no OS-provided caching in that case either.

Both should share the same code.

**Comment by danderson** [comment]
Why is there no OS-provided caching in the UserDial case? Windows, macOS and modern linux all provide an OS-level cache afaik, so the only stuff leftover is older linux?

**Comment by DentonGentry** [comment]
This would be helpful to cache `log.tailscale.io`, which is frequently looked up.

**Comment by ignoramous** [comment]
> From an internal discussion recently: "we can do a proper in-process DNS cache now! (I was waiting for Go 1.19)"

What's special about Go1.19? The [maphash](https://pkg.go.dev/hash/maphash).

**Comment by bradfitz** [comment]
No, not maphash. The DNS resolver work I did for windows. The DNS resolver dialer not works on all platforms so it's possible to intercept all Go DNS client usage and change it.

**Comment by DentonGentry** [comment]
Related: https://github.com/tailscale/tailscale/issues/5161

**Comment by sedlund** [comment]
would this allow two lan hosts to still be able to resolve each other when the internet goes down?
